linker/Linker.Steps/MarkStep.cs

```diff
@@ -1587,5 +1597,143 @@ protected virtual void MarkInterfaceImplementation (TypeDefinition type, Interfa
 			MarkCustomAttributes (iface);
 			MarkType (iface.InterfaceType);
 		}
+
+
+		void MarkSomethingUsedViaReflection (string reflectionMethod, Action<Collection<Instruction>, string, TypeDefinition, BindingFlags> markMethod, Collection<Instruction> instructions)
+		{
+			for (var i = 0; i < instructions.Count; i++) {
+				var instruction = instructions [i];
+				if (instruction.OpCode != OpCodes.Call && instruction.OpCode != OpCodes.Callvirt)
+					continue;
+
+				var methodBeingCalled = instruction.Operand as MethodReference;
+				if (methodBeingCalled == null || methodBeingCalled.DeclaringType.Name != "Type" || methodBeingCalled.DeclaringType.Namespace != "System")
+					continue;
+
+				if (methodBeingCalled.Name != reflectionMethod)
+					continue;
+
+				_context.Tracer.Push ($"Reflection-{methodBeingCalled}");
+				var nameOfThingUsedViaReflection = OperandOfNearestInstructionBefore<string> (i, OpCodes.Ldstr, instructions);
+				var bindingFlags = (BindingFlags) OperandOfNearestInstructionBefore<sbyte> (i, OpCodes.Ldc_I4_S, instructions);
+
+				// There might be more than one ldtoken opcode above the call in the IL stream. Be conservative and check all of
+				// the types which were loaded for the method being used.
+				var declaringTypesOfThingInvokedViaReflection = OperandsOfInstructionsBefore (i, OpCodes.Ldtoken, instructions);
+				foreach (var declaringTypeOfThingInvokedViaReflection in declaringTypesOfThingInvokedViaReflection) {
+					var typeDefinition = declaringTypeOfThingInvokedViaReflection?.Resolve ();
+					if (typeDefinition != null)
+						markMethod (instructions, nameOfThingUsedViaReflection, typeDefinition, bindingFlags);
+				}
+				_context.Tracer.Pop ();
+			}
+		}
+
+		void MarkConstructorsUsedViaReflection (Collection<Instruction> instructions, string unused, TypeDefinition declaringType, BindingFlags bindingFlags)
+		{
+			foreach (var method in declaringType.Methods) {
+				if ((bindingFlags == BindingFlags.Default || bindingFlags.HasFlag (BindingFlags.Public) == method.IsPublic) && method.Name == ".ctor")
+					MarkMethod (method);
+			}
+		}
+
+		void MarkMethodsUsedViaReflection (Collection<Instruction> instructions, string name, TypeDefinition declaringType, BindingFlags bindingFlags)
+		{
+			if (name != null) {
+				foreach (var method in declaringType.Methods)
+				{
+					if ((bindingFlags == BindingFlags.Default || bindingFlags.HasFlag (BindingFlags.Public) == method.IsPublic && bindingFlags.HasFlag(BindingFlags.Static) == method.IsStatic)
```

Comment:
Don't use HasFlags it's slow

Reply:
Why is `HasFlag` slow? I much prefer it to manual bitwise operations, which can be tricky to get correct in my experience.

Reply:
Ok, after some more discussion, I'll not use `HasFlag` and roll my own instead.